### PR TITLE
Add counters to profiler (for CFG spill and reload instructions)

### PR DESCRIPTION
**Pull request by milan-tom**
This PR introduces the ability to measure certain values as counters during passes of the compiler.

Options for API:
1. Existing `Profile.record` function with optional `?counter_f` argument
    - This function is used in many files so changing the signature could potentially mean implementations 
      no longer match interfaces.
2. New `Profile.record_with_counters` function
    - Shows intent better and gives hint about purpose of `couter_f` argument.
    - Allows providing specific documentation for this function.
    - New function so won't break any other files.
    - Intertwines logic for timing/memory profiling with counter profiling (these operate differently as 
      timing and memory are difference measures between the start and end of the stage and have a fixed 
      number of measures whereas counters are purely based on the result of a stage and there may be 
      many counters).
3. Completely separate function
    - No intertwining of logic for timing/memory profiling with counter profiling
    - May lead to code duplication (many aspects in common between these different types of profiling)
    - Harder to print all profile columns for `-dprofile` option (need to manage integration of the two profiling 
      mechanisms)

The current implementation uses option 2.

### CODE REVIEW

**Review by xclerc — COMMENTED** [review]
I think it would be worthwhile to also
amend `/ocaml/driver/compenv.ml`
(in particular the `read_one_param`
function), so that the new behavior
can also be requested through the
`OCAMLRUNPARAM` environment
variable.

**Comment by xclerc** [comment]
(@gretay-js will also review)

File: ocaml/utils/profile.ml [hunk 1]
```diff
@@ -16,26 +16,53 @@
...
 type file = string
+type granularity = File_level | Function_level
+
+module Int = Misc.Stdlib.Int
+module String = Misc.Stdlib.String
+
+module Counters = struct
+  type t = int String.Map.t
+
+  let get name t = String.Map.find name t
```
[the remark this answers is not in the document]
Comment (milan-tom) [reply ocaml/utils/profile.ml:27]: A default value of 0 seems sensible.

File: backend/asmgen.ml [hunk 2]
```diff
@@ -448,7 +461,11 @@ let compile_phrases ~ppf_dump ps =
           if !dump_cmm then fprintf ppf_dump "%a@." Printcmm.phrase p;
           match p with
           | Cfunction fd ->
-            compile_fundecl ~ppf_dump ~funcnames fd;
+            let compile_fundec = compile_fundecl ~ppf_dump ~funcnames in
+            let profiled_compile_fundec = match !profile_granularity with
+            | Function_level -> Profile.record ~accumulate:true fd.fun_name.sym_name compile_fundec
+            | _ -> compile_fundec (* Don't profile if -dfunc-level not selected*)
```
Comment (xclerc) [inline backend/asmgen.ml:467]:
```suggestion
            | File_level -> compile_fundec (* Don't profile if -dfunc-level not selected*)
```
(It is often better to avoid catch-all pattern,
to be notified by the compiler when new
constructors are added to a type.)

Comment (milan-tom) [reply backend/asmgen.ml]: In this case, would it be okay since this part of the code should only be executed if the granularity is `Function_level`? So, even if new granularities are added theoretically this should not change (unless we add a more fine-grained granularity than `Function_level` which seems unlikely).

**Comment by milan-tom** [comment]
> I think it would be worthwhile to also amend `/ocaml/driver/compenv.ml` (in particular the `read_one_param` function), so that the new behavior can also be requested through the `OCAMLRUNPARAM` environment variable.

Done.

**Comment by xclerc** [comment]
Thanks - there was a typo in my comment,
the environment variable is actually named
`OCAMLPARAM` (no `RUN`).

File: backend/asmgen.ml (hunk 2 above)
Comment (xclerc) [reply backend/asmgen.ml]: Generally, we prefer to have to perform
an obvious code update rather than take
the risk of missing an update. Warning 4
is enabled in most files in the backend,
but is unfortunately still disabled in
`Asmgen`.

**Comment by milan-tom** [comment]
> Thanks - there was a typo in my comment, the environment variable is actually named `OCAMLPARAM` (no `RUN`).

Corrected in commit name.

File: ocaml/utils/clflags.mli [hunk 3]
```diff
@@ -59,6 +59,9 @@ module Libloc : sig
   }
 end
 
+type profile_column = [ `Time | `Alloc | `Top_heap | `Abs_top_heap | `Counters ]
```
Comment (gretay-js) [inline ocaml/utils/clflags.mli:62]: Why did `column` move here from `Profile` ?

File: backend/asmgen.ml [hunk 4]
```diff
@@ -448,7 +464,12 @@ let compile_phrases ~ppf_dump ps =
           if !dump_cmm then fprintf ppf_dump "%a@." Printcmm.phrase p;
           match p with
           | Cfunction fd ->
-            compile_fundecl ~ppf_dump ~funcnames fd;
+            let compile_fundec = compile_fundecl ~ppf_dump ~funcnames in
+            (* Only profile if function level granularity selected*)
+            let profiled_compile_fundec = match !profile_granularity with
+            | Function_level -> Profile.record ~accumulate:true fd.fun_name.sym_name compile_fundec
+            | File_level -> compile_fundec
+            in profiled_compile_fundec fd;
```
Comment (gretay-js) [inline backend/asmgen.ml:472]: This is a bit hard to read, maybe just improve formatting or don't let bind `profiled_compile_fundec`.

File: backend/asmgen.ml [hunk 5]
```diff
@@ -348,12 +348,28 @@ let compile_fundecl ~ppf_dump ~funcnames fd_cmm =
...
         in
         cfg
-        ++ begin match regalloc with
-          | GI -> Profile.record ~accumulate:true "cfg_gi" Regalloc_gi.run
-          | IRC -> Profile.record ~accumulate:true "cfg_irc" Regalloc_irc.run
-          | LS -> Profile.record ~accumulate:true "cfg_ls" Regalloc_ls.run
-          | Upstream -> assert false
-        end
+        ++
+        (
```
Comment (gretay-js) [inline backend/asmgen.ml:352]: This part of `Asmgen` is already hard to follow to get an overview of the pipeline. Is it possible to factor out counters into a separate function?

File: ocaml/utils/clflags.mli (hunk 3 above)
Comment (milan-tom) [reply ocaml/utils/clflags.mli:62]: `Profile` needed to know what command line arguments were set to determine whether to compute the counter function or not. So, `column` had to move here to avoid creating a circular dependency between `Profile` and `Clflags`.

File: backend/asmgen.ml (hunk 4 above)
Comment (milan-tom) [reply backend/asmgen.ml]: Replaced with [this](https://github.com/ocaml-flambda/flambda-backend/blob/177ba48b5fa25f618ecedc49e05f7d3790dc4626/backend/asmgen.ml#L468).

**Review by xclerc — APPROVED** [review]
If you can confirm that the output
of `-dtimings` is unaffected when
`-dcounters` is not passed, I think
it is good to go.

(@gretay-js also suggested off-line
it would be useful to give an
example of the new output in the
description of the pull request.)
